Fix for XSLT extension methods format-date & format-time

Maximys opened the pull request:
This is fix for https://github.com/dotnet/runtime/issues/93189 .
I create it as draft, because does not sure that everythink work fine

Maximys committed:
fix(#93189): Fix invalid test.

Maximys committed:
refactor(#93189): Changing tabs to spaces.

Maximys committed:
feat(#93189): Adding test case for check working with fractional.

Maximys committed:
refactor(#93189): Replacing some "InlineDataAttribute" by "MemberDataAttribute".

Maximys committed:
refactor(#93189): Refactoring some tests.

Maximys committed:
refactor(#93189): Extracting "XsdDateTimeParser" and it's environments to separate files.

Maximys committed:
refactor(#93189): Refactoring of "XsdDateTimeParser".

Maximys committed:
refactor(#93189): Renaming "System.Xml.Schema.DateAndTime.Parsers.XsdDateTimeParser" to "System.Xml.Schema.DateAndTime.Converters.DateAndTimeConverter" and change it type from struct to static class.

Maximys commented:
@dotnet/area-system-xml , pay attention, please, that current main brunch of .NET Runtime contains some invalid test cases, which was fixed by https://github.com/dotnet/runtime/pull/114200/commits/c6003c7e4abd5e98932dddb024847f4f397d981e

lilinus commented:
Generally unrelated changes should not be mixed in the same PR (i.e. refactors and bugfixing) according to the [PR guide](https://github.com/dotnet/runtime/blob/main/docs/workflow/ci/pr-guide.md).

Maximys commented:
> Generally, unrelated changes should not be mixed in the same PR (i.e. refactors and bugfixing) according to the [PR guide](https://github.com/dotnet/runtime/blob/main/docs/workflow/ci/pr-guide.md).

Thanks @lilinus, but I think you're not quite right. Most of my refactoring linked with files, which should be touched for fix current bug. May be only tests can be untouched and XsdDateTime with XsdDuration can be left in the previous folder and namespace, but I still think, that this changes make current code a little more readable.

Maximys commented:
@jeffhandley , @dotnet/area-system-xml do you have any feedback?

krwq commented:
@Maximys sorry for the delay. Note this code is in maintenance and therefore the changes should be very minimal and only fix the regression. I think test code looks good for the most part.

Maximys committed:
fix(#93189): Fix functionality of "ms:format-date" function.

Maximys committed:
fix(#93189): Fix functionality of "ms:format-time" function.

Maximys commented:
@krwg, I hope no one will argue that the code quality in System.Private.Xml is not very high; I would even say that in some places it resembles Augean stables.
In my work, I always try to follow the principles from [Code Complete: A Practical Handbook of Software Construction, Second Edition](https://isbnsearch.org/isbn/9780735619678) and the [Boy Scout Rule](https://www.oreilly.com/library/view/97-things-every/9780596809515/ch08.html). Additionally, I always separate commits for refactoring, adding new features, and fixing bugs, using the prefixes refactor, feat, and fix.
The new parser you’re referring to was introduced in commit [95ff95a35916f2084978468628a7cf5f2aab473f](https://github.com/dotnet/runtime/pull/114200/commits/95ff95a35916f2084978468628a7cf5f2aab473f), which also renamed Parser to XsdDateTimeParser. After that, I performed a small refactor in commit [dea09ad1e8f74ea1e794373899800a47ffa8d409](https://github.com/dotnet/runtime/pull/114200/commits/dea09ad1e8f74ea1e794373899800a47ffa8d409). Finally, I fixed the bug in commits [7a30df29b886966b06e920c3a3ba1bef4f48e9f0](https://github.com/dotnet/runtime/pull/114200/commits/7a30df29b886966b06e920c3a3ba1bef4f48e9f0) and [6bdbe81e54776cbc1bb8f47f695ecebcedfb0686](https://github.com/dotnet/runtime/pull/114200/commits/6bdbe81e54776cbc1bb8f47f695ecebcedfb0686).

krwq commented:
@Maximys this change should be very minimal for it to get a chance to make it to .NET 10. As is the earliest is .NET 11 as we consider the amount of changes too risky especially for maintenance package.

I understand that you've put stuff in the separate commits - those are still separate concerns from my perspective and we currently squash every single PR in the end so these won't be visible in case of regression.

Please do investigate if we can simplify the parser - can some existing code be re-used? Current logic might be correct but it's hard to reason about - can you point to pieces of spec you're implementing and write comment why existing APIs can't be used? Can we make some shortcuts to improve readability? (i.e. reduce 80% of code while making most of the scenarios from the spec still work)

I want to emphasize that this package is in maintenance mode - this means we only take low-risk or high impact fixes. This one is considered only because it was a regression from one of the previous releases but the amount of changes here is much greater than original PR which regressed it and touches on the pieces of code which don't seem immediately obvious why they're related and adds a new parser and therefore it's not considered low-risk anymore.

As for quality - take into consideration this library is very old (20+ years?), has many users and it's more than 3MB of source code - while I do agree many pieces are not well written some language constructs and design patterns didn't exist when this code was written. You cannot compare quality without taking that into consideration - every single change you make here can potentially break someone and therefore concerns in this PR. Note that I'm supportive with test refactoring changes - they are little risk from my perspective because they don't touch the product.
